runtime/src/iree-amd-aie/aie_runtime/iree_aie_router.cc

```diff
@@ -730,6 +730,99 @@ bool existsPathToDest(const SwitchSettings &settings, TileLoc currTile,
+    for (int i = 0; i < 5; i++) {
+      if (mask[i] == -1) {
+        mask[i] = id >> i & 0x1;
+      } else if (mask[i] != (id >> i & 0x1)) {
+        // Found bit difference --> mark as "don't care".
+        mask[i] = 2;
+      }
+    }
+  }
+  int maskValue = 0;
+  for (int i = 4; i >= 0; i--) {
+    if (mask[i] == 2) {
+      // Don't care.
+      mask[i] = 0;
+    } else {
+      mask[i] = 1;
+    }
+    maskValue = (maskValue << 1) + mask[i];
+  }
+  for (int id : group) slaveMasks[PhysPortAndID{slavePort, id}] = maskValue;
+}
+
+/// Sort groups by their size in ascending order. A smaller group size can
+/// represent a stricter `packet_rule`, which should be placed first to prevent
+/// other broader (less strict) rules from matching unintended IDs.
+///
+/// Example:
+/// Consider two slave groups, A and B, that share the same `physPort`
+/// (i.e., the same `tileLoc`, `bundle`, and `channel`). These groups
+/// will later be merged into a single `packet_rules` operation,
+/// where each group contributes a `packet_rule` entry. The order
+/// of these entries is critical because the first matching `packet_rule`
+/// takes precedence.
+///
+/// - `Group A` contains IDs `{0x3, 0x4, 0x5}` with `mask = 0x18`,
+///   and defines a `packet_rule`: `(ID & 0x18) == 0x00`.
+/// - `Group B` contains ID `{0x2}` with `mask = 0x1F`,
+///   and defines a `packet_rule`: `(ID & 0x1F) == 0x02`.
+///
+/// In this case, `Group B`'s `packet_rule` must precede `Group A`'s
+/// within the `packet_rules` operation. Otherwise, ID `0x02`
+/// would incorrectly match `(ID & 0x18) == 0x00`, leading to incorrect
+/// behavior.
+void sortGroupsBySize(std::vector<std::set<int>> &groups) {
+  auto sortBySize = [](auto &lhs, auto &rhs) {
+    if (lhs.size() != rhs.size()) return lhs.size() < rhs.size();
+    return lhs < rhs;
+  };
+  std::sort(groups.begin(), groups.end(), sortBySize);
+}
+
+/// Verifies the correctness of ID groupings before putting them into a single
+/// `packet_rules` set. Each group contributes a `packet_rule` entry, and this
+/// function checks if any ID in a later group incorrectly matches a preceding
+/// group's masked ID.
+///
+/// Example:
+/// Consider three groups: A, B, and C.
+/// - `Group A`: Contains IDs `{0x0}` with `mask = 0x1F`.
+///   - Packet rule: `(ID & 0x1F) ?= (0x0 & 0x1F)`.
+/// - `Group B`: Contains IDs `{0x6, 0x7}` with `mask = 0x18`.
+///   - Packet rule: `(ID & 0x18) ?= (0x6 & 0x18)`.
+/// - `Group C`: Contains IDs `{0x1, 0x2, 0x3, 0x4, 0x5}` with `mask = 0x18`.
+///   - Packet rule: `(ID & 0x18) ?= (0x1 & 0x18)`.
+///
+/// ID `0x1` belongs to `Group C`, however, due to the limitation of masking, it
+/// matches both `Group B` and `C`'s rules. Since `Group B` precedes `Group C`,
+/// and `packet_rule` entries are evaluated in order, the function returns
+/// `false` to indicate an invalid grouping.
+bool verifyGroupsByMask(PhysPort slavePort,
+                        const std::vector<std::set<int>> &groups,
```

Comment:
```suggestion
                        const std::vector<std::set<uint32_t>> &groupsPktIds,
```
Use `uint32_t` as it's an ID, usually never negative, and it's used for bitwise operations.